kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/CollectionMatchers.kt

```diff
@@ -176,41 +175,77 @@ fun <T> matchEach(expected: List<T>, asserter: (T, T) -> Unit): Matcher<Collecti
       }
    })
 
+private sealed interface MatchEachProblem {
+   val atIndex: Int
+   val shortMessage: String?
+   val longMessage: String?
+
+   data class SizeMismatch(override val atIndex: Int, val problem: String) : MatchEachProblem {
+      override val shortMessage: String
+         get() = problem
+      override val longMessage: String
+         get() = problem
+   }
+
+   data class Thrown(override val atIndex: Int, val problem: Throwable) : MatchEachProblem {
+      override val shortMessage: String?
+         get() = problem.message
+      override val longMessage: String?
+         get() = problem.stackTraceToString()
+   }
+}
+
 /**
  * Asserts that each element in the collection matches its corresponding matcher in [assertions].
  * Elements will be compared sequentially in the order given by the iterators of the collections.
  */
 fun <T> matchEach(assertions: List<(T) -> Unit>): Matcher<Collection<T>?> = neverNullMatcher { actual ->
-   data class MatchEachProblem(val atIndex: Int, val problem: String?)
-
    val problems = errorCollector.runWithMode(ErrorCollectionMode.Hard) {
       actual.mapIndexedNotNull { index, element ->
          if (index !in assertions.indices) {
-            MatchEachProblem(
+            MatchEachProblem.SizeMismatch(
                index,
                "Element has no corresponding assertion. Only ${assertions.size} assertions provided"
             )
          } else {
             runCatching {
                assertions[index](element)
             }.exceptionOrNull()?.let { exception ->
-               MatchEachProblem(index, exception.message)
+               MatchEachProblem.Thrown(index, exception)
             }
          }
       }
    } + (actual.size until assertions.size).map {
-      MatchEachProblem(
+      MatchEachProblem.SizeMismatch(
          it,
          "No actual element for assertion at index $it"
       )
    }
 
-   MatcherResult(
-      problems.isEmpty(),
-      {
-         "Expected each element to pass its assertion, but found issues at indexes: [${problems.joinToString { it.atIndex.toString() }}]\n\n" +
-            problems.joinToString(separator = "\n") { "${it.atIndex} => ${it.problem}" }
-      },
-      { "Expected some element to fail its assertion, but all passed." },
-   )
+   when (problems.size) {
+      0 -> MatcherResult(
+         true,
+         { "" },
+         { "Expected some element to fail its assertion, but all passed." },
+      )
+
+      1 -> {
+         val problem = problems.first()
+         MatcherResultWithError(
+            (problem as? MatchEachProblem.Thrown)?.problem,
+            false,
+            { "Expected each element to pass its assertion, but found issues at indexes: [${problem.atIndex}}]\n\n${problem.atIndex} => ${problem.shortMessage}" },
```

Comment:
Wouldn't it make more sense to use the long message when there's just the single failure, and shortMessage when there's multiple?

Reply:
No, that's the point of the change. I tried to add the kdocs to the interface to make it more understandable.
Long message contains stacktrace, so it is suitable for use if the stacktrace would otherwise be lost.
On the other hand, short message can be used if the stack trace is preserved via `MatcherResultWithError` - for a single failure.

Reply:
Okay, that makes sense. I think I still need to see how it changes the output of tests to fully understand the context of the changes.